File: 1190/solution.ts
function reverseParentheses(s: string): string {
  const parens = new Map<number, number>();
  const stack: Array<number> = [];
  for (let i = 0; i < s.length; i++) {
    if (s.charAt(i) == "(") {
      stack.push(i);
    } else if (s.charAt(i) == ")") {
      const j = stack.pop()!;
      parens.set(i, j);
      parens.set(j, i);
    }
  }
  let direction = 1;
  let i = 0;
  const res: Array<string> = [];
  while (i >= 0 && i < s.length) {
    if (parens.has(i)) {
      direction *= -1;
      i = parens.get(i)! + direction;
    } else {
      res.push(s[i]);
      i += direction;
    }
  }
  return res.join("");
}

File: 1190/solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public String reverseParentheses(String s) {
        Map<Integer, Integer> parens = new HashMap<>();
        List<Integer> stack = new ArrayList<>();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.add(i);
            } else if (s.charAt(i) == ')') {
                int j = stack.removeLast();
                parens.put(i, j);
                parens.put(j, i);
            }
        }
        int direction = 1;
        int i = 0;
        StringBuilder res = new StringBuilder();
        while (i >= 0 && i < s.length()) {
            if (parens.containsKey(i)) {
                direction *= -1;
                i = parens.get(i) + direction;
            } else {
                res.append(s.charAt(i));
                i += direction;
            }
        }
        return res.toString();
    }
}

// time O(n)
// space O(n)